clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -291,14 +294,28 @@ private KafkaProducer(ProducerConfig config, Serializer<K> keySerializer, Serial
             } else {
                 this.requestTimeoutMs = config.getInt(ProducerConfig.REQUEST_TIMEOUT_MS_CONFIG);
             }
+            boolean idempotenceEnabled = config.getBoolean(ProducerConfig.IDEMPOTENCE_ENABLED_CONFIG);
+
+            if (idempotenceEnabled) {
+                this.transactionState = new TransactionState(idempotenceEnabled);
+                if (config.getInt(ProducerConfig.RETRIES_CONFIG) == 0) {
```

Comment:
I think it's best to throw an exception if the producer isn't configured to be idempotent. Maybe the user doesn't understand the requirements for idempotence, and if they did, may not want to use it? For the initial period especially, I think it is best to set the expectation explicitly.